src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,298 @@
+
+        $cid = array_shift($this->cids);
+        if(Coroutine::exists($cid)) {
+            Coroutine::resume($cid);
+        } else {
+            $this->shiftCoroutine();
+        }
+
+        return $cid;
+    }
+
+    private function pushCoroutine(): mixed
+    {
+        $cid = Coroutine::getCid();
+        $this->cids[] = $cid;
+        Coroutine::yield();
+        return $cid;
+    }
+
+    private function createPromise(): DummyPromise
+    {
+        return new DummyPromise(function () {
+            $this->shiftCoroutine();
+        });
+    }
+
+    private function resolve($value)
+    {
+        if($value instanceof DummyPromise && !$value->isPending()) {
+            if($value->getException() !== null) {
+                throw $value->getException();
+            }
+
+            $value = $this->resolve($value->getValue());
+        }
+
+        if(is_array($value)) {
+            foreach ($value as $k => $v) {
+                $value[$k] = $this->resolve($v);
+            }
+        }
+
+        return $value;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function wait(Promise $promise): mixed
+    {
+        $promise = DummyPromise::wrap($promise);
+
+        if(count($this->cids) === 0 && $promise->isPending()) {
+            throw new \Error('Impossible to resolve the promise, no more task to execute..');
+        }
+
+        while ($promise->isPending()) {
+            $this->shiftCoroutine();
+        }
+
+        return $this->resolve($promise);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function async(Generator $generator): Promise
+    {
+        $fnWrapGenerator = function (Generator $generator, $deferred) use (&$fnWrapGenerator) {
+            Coroutine::create(function () use ($generator, $deferred, $fnWrapGenerator) {
+                try {
+                    while ($generator->valid()) {
+                        $promise = $generator->current();
+                        if(!$promise instanceof DummyPromise) {
+                            throw new \Error('Asynchronous function is yielding a ['.gettype($promise).'] instead of a Promise.');
+                        }
+                        $this->pushCoroutine();
+                        $generator->send($this->resolve($promise));
+                    }
+
+                    $deferred->resolve($generator->getReturn());
+                } catch (Throwable $exception) {
+                    try {
+                        $generator->throw($exception);
+                        $fnWrapGenerator($generator, $deferred);
+                    } catch (\Throwable $throwable) {
+                        $deferred->reject($throwable);
+                    }
+                }
+            });
+        };
+
+        $deferred = $this->createPromise();
+        $fnWrapGenerator($generator, $deferred);
+
+        return $deferred;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function promiseAll(Promise ...$promises): Promise
```

Comment:
Can't we use a [`WaitGroup`°(https://www.swoole.co.uk/docs/modules/swoole-coroutine-waitgroup) ? 🤔